neuraxle/pipeline.py

```diff
@@ -55,13 +60,73 @@ def inverse_transform(self, processed_outputs):
         return self.inverse_transform_processed_outputs(processed_outputs)
 
 
+class PipelineRepository(ABC):
+    def __init__(self, cache_folder, pipeline_cache_list_file_name='pipeline_cache_list.txt'):
+        self.pipeline_cache_list_file_name = pipeline_cache_list_file_name
+        self.cache_folder = cache_folder
+
+    def save(self, pipeline: 'Pipeline', data_container: DataContainer):
+        pipeline_cache_folder = os.path.join(self.cache_folder, pipeline.name)
+
+        if not os.path.exists(pipeline_cache_folder):
+            os.makedirs(pipeline_cache_folder)
+
+        with open(os.path.join(pipeline_cache_folder, self.pipeline_cache_list_file_name), mode='a') as file:
+            next_cached_pipeline_path = self._create_next_cached_pipeline_path(pipeline_cache_folder)
+
+            file.write(str(data_container))
+            file.write(next_cached_pipeline_path)
+
+            dump(pipeline, next_cached_pipeline_path)
```

Comment:
I wouldn't use joblib's `dump` in an abstract class as done here. Like the idea of the `Hashers` we had, we need a way to pass an object to a base step which knows how to serialize it ("it" is the base step here) and deserialize it given it's given serialization subfolder path. For instance, TensorFlow, Keras or PyTorch models may need some special form of serializer adapted for them. Perhaps a visitor pattern as I suggest in the `def save` of the BaseStep? 

Any better idea?

Reply:
it is not inside an abstract class.. it is inside a concrete class... see next commit